cilium/CFP-30984-dns-proxy-ha.md

```diff
@@ -0,0 +1,231 @@
+
+#### Option 1a: gRPC method
+
+Get the endpoint ID and identity for a given IP address by making a gRPC call to cilium agent.
+
+##### Pros
+
+* Avoid interacting with low-level details. Simpler to maintain and stay upto date with any datapath changes.
+* All data can be cached in-memory, in an event of agent being unavailable, SDP can lookup from in-memory cache
+
+##### Cons
+
+* Reliance on gRPC call in the hot-path
+* Does not cache the data for endpoints that never made a DNS query.
+* In an event where SDP restarts when agent's gRPC service is unavailable, all cached state is lost and SDP cannot translate IP to endpoint ID or identity.
+
+#### Option 1b: Listen to ipcache updates via grpc
+
+Similar to envoy, SDP can listen to ipcache updates and maintain a local cache of IP to endpoint ID mappings via grpc. This changes the way DNS proxy currently behaves i.e pull based model to push based model.
+
+##### Pros
+
+* Simpler to maintain and stay upto date with any new ip<>identity changes.
+* All data can be cached in-memory, in an event of agent being unavailable, SDP can lookup from in-memory cache
+
+##### Cons
+
+* Need a mechanism for reconciling the cache in case of SDP restarts.
+* SDP will be aware of all the endpoints data, which might not be needed for DNS proxy.
+
+#### Option 2: Read from bpf maps
+
+Read mappings from bpf maps `cilium_lxc` for endpoint ID and `cilium_ipcache` for identity.
+
+##### Pros
+
+* Reduced resource utilization in agent since agent doesn't need to process rpc calls in hotpath
+* IP to endpoint ID or identity translation does not rely on agent availability.
+* DNS can be resolved even when agent is down.
+
+##### Cons
+
+* Low level interactions from components external to cilium agent.
+
+### Flag to disable built-in proxy
+
+Adding a flag in cilium agent to disable built-in dns proxy allows for de-coupling of SDP's lifecycle from agent. The current implementation of cilium agent depends on the built-in dns proxy for restoring the endpoint for DNS rules. With the flag enabled we will need to refactor to remove that dependency. Cilium agent can retrieve the DNS rules from in memory policy representation (agent already parses this information from ep_config.json on startup).
+
+#### Pros
+
+* Allows for de-coupling upgrade cycles of SDP and cilium agent.
+* Allows any delegated DNS proxy to be plugged in.
+* Reduces the resource utilization in cilium agent.
+
+#### Cons
+
+* Higher latency in DNS resolution as the DNS queries are always handled by a separate process.
+
+### Q: Restoring L7 DNS policy when agent is unavailable
+
+In an event where SDP restarts when agent's gRPC service is unavailable, SDP can read state from `ep_config.json` and restore policy.
```

Comment:
For DNS rules specifically though, we aren't relying on bpf maps right ? ( IIRC, this info isn't stored in any bpf maps right ? )
This just allows SDP to enforce the last known state instead of dropping all DNS requests when agent is down. This sounded to us like a low hanging fruit ( conditional to https://github.com/cilium/cilium/pull/33412 making it in )

Reply:
I just think these aspects are getting ahead of the core problem and trying to solve steps 2 and 3, more niche advanced ordering problems, and it would simplify everything to eliminate the filesystem aspects and focus on the core problems first. 

I don't think it hurts the design or causes additional work to just ... not solve the "bootstrap SDP first" case to begin with. If we want HA, then at least one component must be on all the time. Once both components go down, we've lost HA. If only the agent is up, then SDP restarts, then the agent will still be up to serve the rules. If only SDP is up, then it will keep the rules in memory and doesn't need to restore from disk.

There's a lot of complexity being added by all the filesystem aspects of the CFP. Maybe that complexity is actually worthwhile and we end up agreeing down the line that it should be done this way. Right now though I just think it makes the whole design more complicated than it needs to be in order to solve something that isn't a core problem of the CFP.

Splitting the filesystem aspects out into their own CFP would also make it easier to evaluate just that aspect and for instance whether SDP itself can just store+restore that or whether we need a files-based API agreement between components. It may be low-hanging fruit, but to use the analogy we haven't planted the tree yet :sweat_smile: .